### PR TITLE
Add stdio (`""`/`"-"`) support

### PR DESCRIPTION
**Pull request by sjohannes**
This replaces plain file paths within the program with a `Location { Path(PathBuf), Stdio }` enum, where Stdio represents the `""` or `"-"` command-line argument.

The changes look bigger than they are because I added a bunch of tests (the second commit being all new tests). If you want, I can split up the merge request for easier review.

Feel free to nitpick anything; I'm happy to make any changes. :-)

---

Important note: this adds a dependency on the [`tempfile`](https://crates.io/crates/tempfile) crate.

Because Stdin & Stdout do not implement Seek, which is needed by the codecs, we have two options when reading/writing stdio:

- Buffer the whole image in memory. Not ideal for very large files.
- Use a temporary file. This is the approach I took for this patch.

Based on the output of strace, looks like ImageMagick also uses temporary files. I think it always creates one when reading from stdin, while for stdout it does this for formats that require seeking, e.g. TIFF.

### CODE REVIEW

File: src/arg_parsers/filename.rs [hunk 1]
```diff
@@ -20,78 +20,123 @@ use crate::{arg_parse_err::ArgParseErr, error::MagickError, wm_err};
...
 #[derive(Debug, Clone, PartialEq)]
 pub struct InputFileArg {
-    pub path: PathBuf,
+    pub location: Location,
     pub format: Option<ImageFormat>,
     pub read_mod: Option<ReadModifier>,
 }
 
 impl InputFileArg {
     pub fn parse(input: &OsStr) -> Result<Self, MagickError> {
```
Comment (sjohannes) [inline src/arg_parsers/filename.rs:48]: I rewrote most this method because adding Location handling to the existing code made it really hard to follow. I think this matches ImageMagick's behaviour but more testing would be great. I also added unit tests for some of the edge cases.

File: src/arg_parsers/filename.rs [hunk 2]
```diff
@@ -306,49 +351,41 @@ fn split_off_bracketed_suffix(input: &OsStr) -> Option<(OsString, OsString)> {
...
-                return None;
-            }
-            let suffix = iter.next()?;
-            Some((
-                OsString::from_wide(suffix), // From std::os::windows::ffi::OsStringExt
-                ImageFormat::from_extension(String::from_utf16(prefix).ok()?)?,
-            ))
-        }
-        #[cfg(not(any(unix, windows, target_os = "wasi")))]
+pub fn parse_path_and_format(input: &OsStr) -> Option<(OsString, ImageFormat)> {
```
Comment (sjohannes) [inline src/arg_parsers/filename.rs:354]: This is just moving everything in the inner function to the outer one, because it turns out the return type is already more convenient for the callers.

**Review by Shnatsel — COMMENTED** [review]
Thank you!

I appreciate the use of temporary files up front. I would probably start just by buffering in memory, which is fine for compressed files like JPEG and PNG but becomes a bit of a problem for uncompressed data.

I'm not thrilled about the dependency sprawl of `tempfile`, and we don't need most of those features, but I couldn't find any lightweight alternatives on crates.io that are still maintained. And I'm not (yet) desperate enough to roll my own `micro-tempfile` crate.

I also appreciate the added tests!

I've left just one nit, otherwise looks good to go.

File: src/arg_parsers/filename.rs [hunk 3]
```diff
@@ -20,78 +20,139 @@ use crate::{arg_parse_err::ArgParseErr, error::MagickError, wm_err};
...
+        // If "jpg:foo.jpg[50x50]" doesn't exist or is (a symlink to) a directory,
+        // the next path tried is "jpg:foo.jpg" (with modifier=50x50),
+        // and then the next one is "foo.jpg" (with format=jpg, modifier=50x50).
+        // If none of those files exist:
+        // - If "foo.jpg" exists as a directory, just return it anyway.
+        //   The decoder will deal with the problem later.
+        // - If "foo.jpg" doesn't exist, return error showing "foo.jpg" as the path,
+        //   not the full original string.
+        // If at any point in this process the remaining path becomes "" or "-",
+        // we return the special Stdio location indicating that the input is from stdin.
```
[the remark this answers is not in the document]
Comment (Shnatsel) [reply src/arg_parsers/filename.rs:75]: looks great, thanks!

**Review by Shnatsel — APPROVED** [review]
